### PR TITLE
Issues/91

### PR DESCRIPTION
**Pull request by mike-gangl**
Added 'citation' text file after subscription and downloader runs.

### CODE REVIEW

File: subscriber/podaac_data_downloader.py [hunk 1]
```diff
@@ -284,6 +284,14 @@ def run(args=None):
     logging.info("Downloaded Files: " + str(success_cnt))
     logging.info("Failed Files:     " + str(failure_cnt))
     logging.info("Skipped Files:    " + str(skip_cnt))
+
+    #create citation file if success > 0
+    if success_cnt > 0:
+        try:
+            pa.create_citation_file(short_name, provider, data_path, token, args.verbose)
+        except:
+            logging.debug("Error generating citation")
```
Comment (frankinspace) [inline subscriber/podaac_data_downloader.py:293]:
```suggestion
        except:
            logging.debug("Error generating citation", exc_info=True)
```
IMO we should never ignore stack traces.

File: subscriber/podaac_data_subscriber.py [hunk 2]
```diff
@@ -314,6 +314,13 @@ def run(args=None):
     logging.info("Downloaded Files: " + str(success_cnt))
     logging.info("Failed Files:     " + str(failure_cnt))
     logging.info("Skipped Files:    " + str(skip_cnt))
+
+    if success_cnt > 0:
+        try:
+            pa.create_citation_file(short_name, provider, data_path, token, args.verbose)
+        except:
+            logging.warning("Error generating citation")
```
Comment (frankinspace) [inline subscriber/podaac_data_subscriber.py:322]:
```suggestion
        except:
            logging.warning("Error generating citation", exc_info=True)
```
Is there a reason you went with warning here but debug in downloader?

**Review by frankinspace — CHANGES_REQUESTED** [review]

**Comment by frankinspace** [comment]
Looks like some of the regression tests need to be updated to take into account an additional file potentially being created in the destination directory.

**Comment by mike-gangl** [comment]
@frankinspace agreed, i should have updated those before the PR. got caught up on something but will make them in a minute.

File: subscriber/podaac_data_downloader.py (hunk 1 above)
Comment (mike-gangl) [reply subscriber/podaac_data_downloader.py]: agreed, i added this in- the difficulty is if a non-PO.DAAC collection is used and it doesn't supply all the bits, i don't want to break the rest of subscriber for a this feature.

Comment (mike-gangl) [reply subscriber/podaac_data_downloader.py]: i added the exc_info=True to the warning message

File: subscriber/podaac_data_subscriber.py (hunk 2 above)
Comment (mike-gangl) [reply subscriber/podaac_data_subscriber.py]: no...

**Comment by frankinspace** [comment]
@mike-gangl I enabled debug logging during regression tests so we can see the logs when there are failures, it looks like it is still random 401 errors from CMR causing the problem.

I also set the max parallel to 2 so it will only run 2 combinations of the matrix in parallel in the hopes that reducing the number of parallel tests we'd get some more consistency (at the expense of the tests taking longer). It looks like that did help a bit but is still not enough.

**Comment by mike-gangl** [comment]
thanks @frankinspace . I've updated the test to look for the data files only given that a citation file _might_ not be created. i figured it wouldn't be created due to a different DAAC not having the required information, but token issues in CMR are another reason here.

File: subscriber/podaac_data_downloader.py [hunk 3]
```diff
@@ -272,7 +272,9 @@ def run(args=None):
                 skip_cnt += 1
                 continue
 
-            urlretrieve(f, output_path)
+            pa.download_file(f,output_path)
+            #urlretrieve(f, output_path)
```
Comment (frankinspace) [inline subscriber/podaac_data_downloader.py:276]: Remove the commented out code

File: subscriber/podaac_access.py [hunk 4]
```diff
@@ -287,6 +290,26 @@ def get_temporal_range(start, end, now):
     raise ValueError("One of start-date or end-date must be specified.")
 
 
+def download_file(remote_file, output_path, retries=3):
+    failed = False
+    for r in range(retries):
```
Comment (frankinspace) [inline subscriber/podaac_access.py:295]: For consistency should probably use `@tenacity.retry`.
Check out https://github.com/podaac/data-subscriber/pull/92/files/5ad8e149843d5499b8601740cbc963aaac0f46f5..78aac3ffb4d6586fbc796935e85e67df41f5b53b#diff-158519578c6b1bf3dc62de68856136d011247d7ed6b085dfb341dfe452d2a1cfR314

**Pull request merged**